crates/biome_js_analyze/src/lint/nursery/use_unified_type_signature.rs

```diff
@@ -0,0 +1,1221 @@
+    TsConstructSignatureTypeMember, TsConstructorSignatureClassMember,
+    TsDeclareFunctionDeclaration, TsDeclareFunctionExportDefaultDeclaration, TsDeclareStatement,
+    TsMethodSignatureClassMember, TsMethodSignatureTypeMember, TsTypeParameters,
+};
+use biome_rowan::{
+    AstNode, AstSeparatedList, BatchMutation, BatchMutationExt, SyntaxResult, TextRange,
+    TriviaPieceKind, chain_trivia_pieces, declare_node_union,
+};
+
+declare_lint_rule! {
+    /// Disallow overload signatures that can be unified into a single signature.
+    ///
+    /// Overload signatures that can be merged into a single signature are redundant and should be avoided.
+    /// This rule helps simplify function signatures by combining overloads by making parameters optional
+    /// and/or using type unions.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```ts,expect_diagnostic
+    /// function f(a: number): void;
+    /// function f(a: string): void;
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// interface I {
+    ///     a(): void;
+    ///     a(x: number): void;
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```ts
+    /// function f(a: number | string): void {}
+    /// ```
+    ///
+    /// ```ts
+    /// interface I {
+    ///     a(x?: number): void;
+    /// }
+    /// ```
+    ///
+    /// Different return types cannot be merged:
+    /// ```ts
+    /// interface I {
+    ///     f(): void;
+    ///     f(x: number): number;
+    /// }
+    /// ```
+    ///
+    /// Different type parameters cannot be merged:
+    /// ```ts
+    /// function f<T extends number>(x: T): void;
+    /// function f<T extends string>(x: T): void;
+    /// function f(x: unknown): void {}
+    /// ```
+    ///
+    pub UseUnifiedTypeSignature {
+        version: "next",
+        name: "useUnifiedTypeSignature",
+        language: "ts",
+        recommended: false,
+        sources: &[RuleSource::EslintTypeScript("unified-signatures")],
+        fix_kind: FixKind::Unsafe,
+    }
+}
+
+impl Rule for UseUnifiedTypeSignature {
+    type Query = Ast<AnyPotentialTsOverloadSignature>;
+    type State = MergeOverloadSignaturesInfo;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+
+        let overload_info = OverloadInfo::from_overload_signature(node)?;
+
+        // Yes, this is a O(n^2) algorithm, but we need to compare all overload signatures
+        // with each other. This shouldn't be a problem in practice,
+        // since the number of overloads is usually relatively small.
+        for i in 0..overload_info.overload_signatures.len() {
+            for j in (i + 1)..overload_info.overload_signatures.len() {
+                let overload1 = &overload_info.overload_signatures[i];
+                let overload2 = &overload_info.overload_signatures[j];
```

Comment:
I would suggest writing it like this, since it will help the compiler eliminate bound checking:

```suggestion
        for (i, overload1) in overload_info.overload_signatures.iter().enumate() {
            for overload2 in &overload_info.overload_signatures[i + 1..] {
```